shaders/src/light_reflections.fs

```diff
@@ -0,0 +1,254 @@
+
+highp mat4 scaleMatrix(const highp float x, const highp float y) {
+    mat4 m = mat4(1.0f);
+    m[0].x = x;
+    m[1].y = y;
+    m[2].z = 1.0f;
+    m[3].w = 1.0f;
+    return m;
+}
+
+/**
+ * Evaluates screen-space reflections, storing the color in Fr.rgb if there's a hit.
+ * r is the desired reflected vector.
+ *
+ * Fr.a is set to a value between [0, 1] representing the "opacity" of the reflection. 1.0f is full
+ * screen-space reflection. Values < 1.0f should be blended with the scene's IBL.
+ *
+ * If there is no hit, Fr is unmodified.
+ */
+void evaluateScreenSpaceReflections(vec3 r, inout vec4 Fr) {
+    vec3 wsRayDirection = r;
+    highp vec3 wsRayStart = shading_position + frameUniforms.ssrBias * wsRayDirection;
+
+    // ray start/end in view space
+    highp vec4 vsRayStart = getViewFromWorldMatrix() * vec4(wsRayStart, 1.0);
+    highp vec4 vsRayDirection = getViewFromWorldMatrix() * vec4(wsRayDirection, 0.0);
+
+    highp vec3 vsOrigin = vsRayStart.xyz;
+    highp vec3 vsDirection = vsRayDirection.xyz;
+    float vsZThickness = frameUniforms.ssrThickness;
+    // TODO: use the actual near plane.
+    const float nearPlaneZ = -0.1f;
+    const float stride = 1.0;
+    // TODO: jitterFraction should be between 0 and 1, but anything < 1 gives banding artifacts.
+    const float jitterFraction = 1.0f;
+    float maxRayTraceDistance = frameUniforms.ssrDistance;
+
+    highp vec2 res = vec2(textureSize(light_structure, 0).xy);
+    highp mat4x4 projectToPixelMatrix =
+        scaleMatrix(res.x, res.y) *
+        frameUniforms.ssrProjectToPixelMatrix;
+
+    highp float maxSteps = float(max(res.x, res.y));
+
+    // Outputs from the traceScreenSpaceRay function.
+    highp vec2 hitPixel;  // not currently used
+    highp vec3 vsHitPoint;
+
+    if (traceScreenSpaceRay(vsOrigin, vsDirection, projectToPixelMatrix, light_structure,
+            vsZThickness, nearPlaneZ, stride, jitterFraction, maxSteps,
+            maxRayTraceDistance, hitPixel, vsHitPoint)) {
+        highp vec4 reprojected = frameUniforms.ssrReprojection * vec4(vsHitPoint, 1.0f);
+        reprojected *= (1.0 / reprojected.w);
+
+        // Compute the screen-space reflection's contribution.
+        // TODO: parameterize fadeRate.
+        const float fadeRate = 12.0f;
+
+        // Fade the reflections out near the edges.
+        vec2 edgeFactor = max(fadeRate * abs(reprojected.xy - 0.5f) - (fadeRate / 2.0f - 1.0f), 0.0f);
+        float edgeFade = saturate(1.0 - dot(edgeFactor, edgeFactor));
+
+        // Fade the reflections out near maxRayTraceDistance.
+        float t = distance(vsOrigin, vsHitPoint) / maxRayTraceDistance;
+        float distFactor = max(fadeRate * (t - 0.5f) - (fadeRate / 2.0f - 1.0f), 0.0f);
+        float distanceFade = saturate(1.0 - (distFactor, distFactor));
+
+        Fr = vec4(textureLod(light_ssr, reprojected.xy, 0.0f).rgb, edgeFade * distanceFade);
```

Comment:
Usually we use premultiplied alpha everywhere.  Maybe it would be better to stick to that convention. This would just change the `mix` at the call site.